File: quanto/library/python/ungroup.py
import torch


@torch.library.impl("quanto_py::ungroup", "default")
def ungroup(grouped: torch.Tensor, axis: int, orig_shape: torch.Size) -> torch.Tensor:
    if grouped.shape == orig_shape:
        return grouped
    if axis == 0:
        # No transposition required, just reshape
        return grouped.reshape(orig_shape)
    group_size = grouped.shape[0] if axis == -1 else grouped.shape[-1]
    axis_dim = orig_shape[axis]
    groups_per_axis = grouped.numel() // axis_dim // group_size
    ungrouped = grouped.reshape(group_size, axis_dim, groups_per_axis)
    # A dual tranposition is required to reorder to (groups_per_axis, group_size, axis_dim)
    ungrouped = ungrouped.transpose(1, 2)
    ungrouped = ungrouped.transpose(0, 1)
    return ungrouped.reshape(orig_shape)

File: quanto/library/ext/cpp/ungroup.cpp
#include "ungroup.h"
#include <torch/extension.h>

torch::Tensor ungroup(torch::Tensor &grouped, int axis, torch::IntArrayRef orig_shape){
    if (grouped.sizes() == orig_shape){
        return grouped;
    }
    if (axis == 0) {
        return torch::reshape(grouped, orig_shape);
    }
    int64_t group_size = (axis == -1) ? grouped.size(0) : grouped.size(-1);
    int64_t axis_dim = (axis == -1) ? orig_shape.back() : orig_shape[axis];
    // Calculate the number of groups per axis
    int64_t groups_per_axis = grouped.numel() / axis_dim / group_size;

    torch::Tensor ungrouped = grouped.reshape({group_size, axis_dim, groups_per_axis});
    ungrouped = ungrouped.transpose(1, 2);
    ungrouped = ungrouped.transpose(0, 1);

    // Reshape to the original shape
    return ungrouped.reshape(orig_shape);
}